pkg/controllers/failuredomainrollout/controller.go

```diff
@@ -0,0 +1,364 @@
+// Copyright 2025 Nutanix. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package failuredomainrollout
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type Reconciler struct {
+	client.Client
+}
+
+func (r *Reconciler) SetupWithManager(
+	mgr ctrl.Manager,
+	options *controller.Options,
+) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.Cluster{}).
+		Watches(
+			&controlplanev1.KubeadmControlPlane{},
+			handler.EnqueueRequestsFromMapFunc(r.kubeadmControlPlaneToCluster),
+			builder.WithPredicates(predicate.ResourceVersionChangedPredicate{}),
```

Comment:
I don't see it's necessary to watch KubeadmControlPlane for this controller. I think this controller should only watch the Cluster.status.